core/src/backend/audio/decoders/adpcm.rs

```diff
@@ -95,108 +39,81 @@ impl<R: Read> AdpcmDecoder<R> {
-            self.left_step_index = self.inner.read::<u16>(6)? as i16;
-            self.left_step = Self::STEP_TABLE[self.left_step_index as usize];
-            if self.is_stereo {
-                self.right_sample = self.inner.read_signed(16)?;
-                self.right_step_index = self.inner.read::<u16>(6)? as i16;
-                self.right_step = Self::STEP_TABLE[self.right_step_index as usize];
+            for channel in &mut self.channels {
+                channel.sample = self.inner.read_signed(16).ok()?;
+                channel.step_index = self.inner.read::<u16>(6).ok()? as i16;
             }
         }
 
         self.sample_num = (self.sample_num + 1) % 4095;
 
-        let data = self.inner.read::<u32>(self.bits_per_sample as u32)? as i32;
-        self.left_step = Self::STEP_TABLE[self.left_step_index as usize];
-
-        // (data + 0.5) * step / 2^(bits_per_sample - 2)
-        // Data is sign-magnitude, NOT two's complement.
-        // TODO(Herschel): Other implementations use some bit-tricks for this.
-        let sign_mask = 1 << (self.bits_per_sample - 1);
-        let magnitude = data & !sign_mask;
-        let delta = (self.decoder)(self.left_step, magnitude);
-
-        self.left_sample = if (data & sign_mask) != 0 {
-            (self.left_sample as i32 - delta as i32).max(i16::MIN.into())
-        } else {
-            (self.left_sample as i32 + delta as i32).min(i16::MAX.into())
-        } as i16;
-
-        self.left_step_index += Self::INDEX_TABLE[self.bits_per_sample - 2][magnitude as usize];
-        if self.left_step_index < 0 {
-            self.left_step_index = 0;
-        } else if self.left_step_index >= Self::STEP_TABLE.len() as i16 {
-            self.left_step_index = Self::STEP_TABLE.len() as i16 - 1;
-        }
-
-        if self.is_stereo {
-            let data = self.inner.read::<u32>(self.bits_per_sample as u32)? as i32;
-            self.right_step = Self::STEP_TABLE[self.right_step_index as usize];
-
+        for channel in &mut self.channels {
+            // `data` is sign-magnitude, NOT two's complement.
+            let data = self.inner.read::<u32>(self.bits_per_sample as u32).ok()?;
             let sign_mask = 1 << (self.bits_per_sample - 1);
             let magnitude = data & !sign_mask;
-            let delta = (self.decoder)(self.right_step, magnitude);
 
-            self.right_sample = if (data & sign_mask) != 0 {
-                (self.right_sample as i32 - delta as i32).max(i16::MIN.into())
+            // delta = (data + 0.5) * step / 2^(bits_per_sample - 2)
+            let mut delta = 0;
+            let mut step = STEP_TABLE[channel.step_index as usize];
+            let mut k = 1 << (self.bits_per_sample - 2);
+            loop {
+                if (data & k) != 0 {
+                    delta += step;
+                }
+
+                step >>= 1;
+
+                k >>= 1;
+                if k == 0 {
+                    break;
+                }
+            }
```

Comment:
I think I'd prefer doing either the function pointers or a match here (maybe with a macro to avoid repeated code) -- since this is an inner mixing loop we should care a little bit about making it optimized.

Reply:
Reverted it for now as I'm not too expert in Rust macros. I think we should somehow run benchmarks on this to get maximum performance.